File: Lidar_python/Lidar_Traitement.py
## Importation des modules ##

import numpy as np
import matplotlib.pyplot as plt
from math import cos,sin,pi

## Définition des fonctions ##

def getData(filePath):
    '''
    Récupére les données envoyés par le Lidar (mettre un chemin absolu pour le chemin du fichier)
    '''

    x = np.array([])
    y = np.array([])

    with open(filePath, 'r') as fr:       #Ouvre le fichier texte
        linesTxt = fr.readlines()
        ptr = 1
        with open('./RPLIDAR2.txt','w') as fw:
            for lineTxt in linesTxt:                        # Enlève les 3 premieres lignes du fichier et rempli dans RPLIDAR2.txt
                if (ptr !=1 and ptr != 2 and ptr != 3):
                    l=lineTxt.split(' ')
                    x = np.append(x,l[0]) 
                    y = np.append(y,l[1])
                    fw.write(lineTxt)
                ptr += 1
            fw.close()
        fr.close()

    A = np.concatenate([[x], [y]], axis=0)

    return A


def changeBase(A):
    '''
    Passage des données en une base (distance, angle) en une base (x, y)
    '''
    X = np.array([])
    Y = np.array([])
    for (dist,angle) in zip(A[0,:],A[1,:]) :
        dist_i=float(dist)
        angle_i=float(angle)
        b1=angle_i*cos(dist_i*pi/180)/1000
        b2=angle_i*sin(dist_i*pi/180)/1000
        X=np.append(X, b1)
        Y=np.append(Y, b2)

    return X, Y


def createLines(X, Y):
    '''
    Crée les lignes à partir du nuage de point contenu dans X, Y et ne prend en compte que les lignes ayant au moins n points
    '''
    n = 2
    k = 0
    d = 0.01        # Rayon max dans lequel se trouvent 2 points d'une même ligne
    

    lines = np.empty((1, 2), int)
    coeffsDir = np.empty((1, 1), int)

    while(k<len(X)-1) :
        flag = True
        line = np.empty((1, 2), int)
        
    
        while(flag and k < len(X)-1) : 
                
            if((X[k+1] - X[k])**2 + (Y[k+1] - Y[k])**2 < d**2) :
                line = np.append(line, [[ X[k+1], Y[k+1] ]], axis=0)
                    
            else :
                flag = False
                
            k += 1
    
        if(len(line) > n+1) :
            
            line = np.delete(line, 0, axis=0)
            
            m = (line[-1, 1] - line[0, 1])/(line[-1, 0] - line[0, 0])
            coeffsDir = np.append(coeffsDir, [[m]], axis=0) 
            
                                
            lines = np.append(lines, [ line[0,:], line[-1,:] ], axis=0)         

    lines = np.delete(lines, 0, axis=0)
    return lines, coeffsDir
    

def reshapeLines(lines, coeffsDir, gamma):
    '''
    Relie les lignes si leur coeffs directeur et leur position sont similaire à gamma près
    '''
    linesReshape = np.empty((1, 2), int)
    coeffsDir = np.delete(coeffsDir, 0, axis=0)

    for i in range(0, len(lines) - 2, 2):

        linesReshape = np.append(linesReshape, [ lines[i,:] ], axis = 0)

        if(abs(coeffsDir[int(i/2)] - coeffsDir[int(i/2 + 1)]) < gamma
        and abs(lines[i, 0] - lines[i+2, 0]) < gamma
        and (lines[i, 1] - lines[i+2, 1]) < gamma):

            linesReshape = np.append(linesReshape, [ lines[i+2,:] ], axis = 0)

        else:

            linesReshape = np.append(linesReshape, [ lines[i+1,:] ], axis = 0)
                
                
    linesReshape = np.delete(linesReshape, 0, axis=0)
    print(np.size(linesReshape))
    return linesReshape



def printLines(lines):
    '''
    Affiche la carte sous forme de lignes
    '''

    for k in range(int(len(lines)/2)):
        print(k)
        print(lines[2*k, 0], lines[2*k+1, 0], lines[2*k, 1], lines[2*k+1, 1])
        plt.plot([ lines[2*i, 0], lines[2*i+1, 0] ], [ lines[2*i, 1], lines[2*i+1, 1] ])

    plt.draw()
    plt.show()
    


## Tests ##
'''
A = getData('Lidar_python/RPLIDAR.txt')

X, Y = changeBase(A)


lines  = createLines(X, Y, 5)[0]
coeffsDir = createLines(X, Y, 5)[1]
linesReshape = reshapeLines(lines, coeffsDir, 0.07)
'''

File: matlab/Lidar_traitement.m
close all;
clear variables;
clc;

%% Variables globales

sigma = 0.01; % Largeur d'une ligne
d = 0.01;      % Rayon max dans lequel se trouvent 2 points d'une même ligne
n = 10;       % Nbs de points minimum dans une ligne
gamma = 0.4; % Difference accepté pour regrouper 2 lignes

%% Récupération des données

%Ouvre le fichier texte
[fid, message]=fopen("RPLIDAR.txt","r");
%enlève les 3 premieres lignes du fichier
for i=1:3
    fgetl(fid);
end

[A,count]=fscanf(fid,'%g%g%g',[3 inf]); %sépare les données en 3 colonnes
A(3,:)=[];%supprime la ligne qualité

%% Changement de bases

X=A(2,:).*cos(A(1,:).*(pi/180))./1000;
Y=A(2,:).*sin(A(1,:).*(pi/180))./1000;

points = [X;Y];

%% Traçage de lignes
k = 1;
lines = [];
coeffsDir = [];

while(k<length(X)-1) 
   flag = true;
   line = [];
   
   while(flag && k<length(X)-1)
       
       [Cx, Cy] = circle(X(k), Y(k), d);
      % plot(Cy, Cx); hold on;
       
       if((X(k+1) - X(k))^2 + (Y(k+1) - Y(k))^2 < d^2)
           line = cat(1, line, [X(k+1), Y(k+1)]);
           
       else
           flag = false;
       end
       
       k = k + 1;
       
   end
   
   if(isempty(line) == false)
       m = (line(end, 2) - line(1, 2))/(line(end, 1) - line(1, 1)); % Coeff directeur de la ligne en cours
       coeffsDir = cat(1, coeffsDir, m);                            % Ajout du coeff dans la matrice lines
       
       lines = cat(1, lines, [line(1,:); line(end,:)]); % Ajout des coors des 2 point de la ligne dans la matrice lines
  
       
       
   end
end

% for i = 2:2:length(lines)
%     for j = 2:2:length(lines)
%        if(i ~= j && abs(coeffsDir(i/2) - coeffsDir(j/2)) < gamma ...
%                && abs(lines(i, 1) - lines(j, 1)) < gamma && abs(lines(i, 2) - lines(j, 2)) < gamma)
%            lines(i,:) = []; 
%            lines(i+1,:) = [];
% 
%        else
%        end
%     end
%    
% end

%% Affichage
for i = 0:length(lines)/2 - 1
    plot(lines(1+2*i:2*(i+1),1), lines(1+2*i:2*(i+1), 2)); hold on;
   
end

status=fclose(fid); %ferme fichier